### PR TITLE
uoa support ipv6 and ipv4

### PR DESCRIPTION
**Pull request by xiangp126**

### CODE REVIEW

File: kmod/uoa/uoa.c [hunk 1]
```diff
@@ -490,29 +518,49 @@ static int uoa_so_get(struct sock *sk, int cmd, void __user *user, int *len)
...
 	/* lookup uap mapping table */
-	um = uoa_map_get(map.saddr, map.daddr, map.sport, map.dport);
+	um = uoa_map_get(map.af, &map.saddr, &map.daddr, map.sport, map.dport);
+
 	if (!um) {
 		if (uoa_debug) {
-			pr_warn("%s: not found: %pI4:%d->%pI4:%d\n", __func__,
-				&map.saddr, ntohs(map.sport),
-				&map.daddr, ntohs(map.dport));
+            if (AF_INET == map.af) {
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: fix indent problem

File: kmod/uoa/uoa.c [hunk 2]
```diff
@@ -624,26 +672,36 @@ static struct uoa_map *uoa_parse_ipopt(unsigned char *optptr, int optlen,
...
-			um->daddr = daddr;
-			um->sport = sport;
-			um->dport = dport;
-
-			memcpy(&um->optuoa, optptr, IPOLEN_UOA_IPV4);
-
-			UOA_STATS_INC(uoa_saved);
-			return um;
-		}
+        if (*optptr == IPOPT_UOA) {
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: fix indent problem

File: kmod/uoa/uoa.c [hunk 3]
```diff
@@ -725,38 +781,66 @@ static struct uoa_map *uoa_opp_rcv(struct iphdr *iph, struct sk_buff *skb)
...
+        }
+        /* re-calc checksum */
+        ip_send_check(iph);
+    } else {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0)
+        /* do as upper ipv4, handle for old kernel version */
+        int payload_len = ntohs(((struct ipv6hdr *)iph)->payload_len);
+        ((struct ipv6hdr *)iph)->payload_len = htons(payload_len - opplen);
+        ((struct ipv6hdr *)iph)->nexthdr = opph->protocol;
+        memmove(iph + iphdrlen, uh, ntohs(uh->len));
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: Does it work if not move the UDP data?

File: kmod/uoa/uoa.c [hunk 4]
```diff
@@ -725,38 +781,66 @@ static struct uoa_map *uoa_opp_rcv(struct iphdr *iph, struct sk_buff *skb)
...
+        skb_set_transport_header(skb, iphdrlen);
+#else
+         ((struct ipv6hdr *)iph)->nexthdr = opph->protocol;
+#endif
+    }
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: fix the indent problem

File: kmod/uoa/uoa.c [hunk 5]
```diff
@@ -824,8 +917,11 @@ static __init int uoa_init(void)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,3,0)
 	err = nf_register_net_hooks(&init_net, uoa_nf_hook_ops,
 				    ARRAY_SIZE(uoa_nf_hook_ops));
+	err = nf_register_net_hooks(&init_net, uoa_nf_hook_ops6,
+				    ARRAY_SIZE(uoa_nf_hook_ops6));
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: Previous `err` is overwritten before processed.

File: kmod/uoa/uoa.c [hunk 6]
```diff
@@ -824,8 +917,11 @@ static __init int uoa_init(void)
...
 #else
 	err = nf_register_hooks(uoa_nf_hook_ops, ARRAY_SIZE(uoa_nf_hook_ops));
+	err = nf_register_hooks(uoa_nf_hook_ops6, ARRAY_SIZE(uoa_nf_hook_ops6));
```
Comment (ywc689) [inline kmod/uoa/uoa.c]: Previous `err` is overwritten before processed.

**Comment by ywc689** [comment]
Both Tab and Space chars are used for indentation in "uoa.c". Please fix the problem.

File: kmod/uoa/uoa.c (hunk 3 above)
Comment (xiangp126) [reply kmod/uoa/uoa.c]: 这是针对不同内核提供了两种方法，都测试过了

File: kmod/uoa/example/opp.c [hunk 7]
```diff
@@ -28,6 +28,8 @@
 #include <netinet/udp.h>
 #include <arpa/inet.h>
 #include "common.h"
+
+#include "inet.h"
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply kmod/uoa/example/opp.c]: removed

**Pull request merged**